File: scripts/check-capture-env.ts
#!/usr/bin/env bun

import { spawnSync } from "node:child_process";
import { existsSync } from "node:fs";
import { resolve } from "node:path";

const ANSI_RE = /\x1b\[[0-?]*[ -/]*[@-~]/g;
const ROOT = resolve(import.meta.dir, "..");
const LOCAL_CHROME_WRAPPER = resolve(ROOT, "scripts", "chrome-with-libs.sh");

function clean(text: string): string {
  return text.replace(ANSI_RE, "").trim();
}

function run(cmd: string[]): ReturnType<typeof spawnSync> {
  return spawnSync(cmd[0], cmd.slice(1), { encoding: "utf8" });
}

function browserCommandArgs(): string[] {
  if (existsSync(LOCAL_CHROME_WRAPPER)) {
    return ["--executable-path", LOCAL_CHROME_WRAPPER];
  }

  return [];
}

const version = run(["agent-browser", "--version"]);

if (version.status !== 0) {
  console.error("agent-browser is not installed or not available on PATH.");
  console.error(clean(version.stderr || version.stdout || ""));
  process.exit(1);
}

const versionText = clean(version.stdout || version.stderr || "");
const smoke = run(["agent-browser", ...browserCommandArgs(), "--session", "capture-preflight", "open", "about:blank"]);

if (smoke.status !== 0) {
  const output = clean(`${smoke.stderr || ""}\n${smoke.stdout || ""}`);
  console.error("Capture preflight failed.");
  console.error(output);

  if (output.includes("libnspr4.so")) {
    console.error(
      "Missing Chrome runtime dependency detected: libnspr4.so. Install the required system libraries before running capture."
    );
    console.error(
      "Typical Debian/Ubuntu fix: sudo apt-get install -y libnspr4 libnss3 libatk-bridge2.0-0 libxkbcommon0 libgtk-3-0"
    );
    console.error(
      "On this repo's Arch-based setup, you can also run: bun run capture:bootstrap-libs"
    );
  }

  console.error(
    "If your environment already has a working browser, you can also point agent-browser at it with --executable-path."
  );
  console.error(versionText);
  process.exit(1);
}

const closeResult = run(["agent-browser", ...browserCommandArgs(), "--session", "capture-preflight", "close"]);
if (closeResult.status !== 0) {
  console.error(clean(closeResult.stderr || closeResult.stdout || ""));
}

console.log(versionText);
console.log("Capture environment ready.");

File: scripts/check_capture_env.py
#!/usr/bin/env python3

from __future__ import annotations

import re
import subprocess
import sys


ANSI_RE = re.compile(r"\x1b\[[0-?]*[ -/]*[@-~]")


def clean(text: str) -> str:
    return ANSI_RE.sub("", text).strip()


def run(cmd: list[str]) -> subprocess.CompletedProcess[str]:
    return subprocess.run(cmd, text=True, capture_output=True)


def main() -> int:
    version = run(["agent-browser", "--version"])
    if version.returncode != 0:
        print(
            "agent-browser is not installed or not available on PATH.", file=sys.stderr
        )
        print(clean(version.stderr or version.stdout), file=sys.stderr)
        return 1

    version_text = clean(version.stdout or version.stderr)

    smoke = run(
        ["agent-browser", "--session", "capture-preflight", "open", "about:blank"]
    )
    if smoke.returncode != 0:
        output = clean((smoke.stderr or "") + "\n" + (smoke.stdout or ""))
        print("Capture preflight failed.", file=sys.stderr)
        print(output, file=sys.stderr)

        if "libnspr4.so" in output:
            print(
                "Missing Chrome runtime dependency detected: libnspr4.so. Install the required system libraries before running capture.",
                file=sys.stderr,
            )
            print(
                "Typical Debian/Ubuntu fix: sudo apt-get install -y libnspr4 libnss3 libatk-bridge2.0-0 libxkbcommon0 libgtk-3-0",
                file=sys.stderr,
            )

        print(
            "If your environment already has a working browser, you can also point agent-browser at it with --executable-path.",
            file=sys.stderr,
        )
        print(version_text, file=sys.stderr)
        return 1

    close_result = run(["agent-browser", "--session", "capture-preflight", "close"])
    if close_result.returncode != 0:
        print(clean(close_result.stderr or close_result.stdout), file=sys.stderr)

    print(version_text)
    print("Capture environment ready.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
